File: programmers/level2/42885.js
//https://school.programmers.co.kr/learn/courses/30/lessons/42885
//구명보트

function solution(people, limit) {
  let answer = 0;

  const temp = [...people];
  temp.sort((a,b)=> b-a)
  let weight = temp.shift();
  answer++;
  let idx = 0;
  while (temp.length > 0) {
    // console.log('temp', temp);
    if (idx === temp.length) {
      answer++;
      idx = 0;
      weight = temp.shift();
      // console.log('----');
    }
    // console.log('weight', weight);
    if (weight + temp[idx] <= limit) {
      const [el] = temp.splice(idx, 1);
      // console.log('el', el);
      weight += el;
      // console.log('add', weight);
    } else {
      idx++;
      // console.log('next', idx);
    }
  }

  return answer;
}

console.log(solution([70, 50, 80, 50], 100)); //3
console.log(solution([70, 80, 50], 100)); //3

File: programmers/python/level2/42885.py
def solution(people, limit):
  answer = 0
  people.sort()
  # print(people)
  in_boat = 0
  need_check= len(people)-1

  while in_boat < need_check:
    print('people[in_boat]',in_boat,people[in_boat])
    print('people[need_check]',need_check,people[need_check])
    if people[need_check] + people[in_boat] <= limit:
      in_boat += 1
      answer += 1
    need_check -= 1
  return len(people)-answer

print(solution([70, 50, 80, 50, 50 ],	100	))
# print(solution([70, 50, 80, 50],	100	))
# print(solution([70, 80, 50],	100	))
# print(solution([50, 50, 50, 60, 70, 80, 40],	100	))
# [70, 50, 80, 50]	100	3
# [70, 80, 50]	100	3
